File: lib/domain/interaction_rule/ports/interaction_rule_detail_output_port.rb
# frozen_string_literal: true

module Domain
  module InteractionRule
    module Ports
      class InteractionRuleDetailOutputPort
        def on_success(detail_dto)
          raise NotImplementedError, "Subclasses must implement on_success"
        end

        def on_failure(error_dto)
          raise NotImplementedError, "Subclasses must implement on_failure"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/interaction_rule/ports/interaction_rule_detail_output_port.rs
use crate::interaction_rule::dtos::InteractionRuleDetailOutput;
use crate::shared::dtos::error::Error;
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;

/// Ruby: `Domain::InteractionRule::Ports::InteractionRuleDetailOutputPort`
pub trait InteractionRuleDetailOutputPort {
    fn on_success(&mut self, dto: InteractionRuleDetailOutput);
    fn on_failure(&mut self, error: DetailFailure);
}

#[derive(Debug, Clone, PartialEq)]
pub enum DetailFailure {
    Policy(PolicyPermissionDenied),
    Error(Error),
}
